src/ORM/EagerLoadedDataList.php

```diff
@@ -0,0 +1,893 @@
+        }
+
+        $this->dataClass = $dataClass;
+        $this->foreignID = $foreignID;
+
+        // many_many relation lists have extra constructor args that don't apply for has_many or non-relations
+        if (is_a($dataListClass, ManyManyThroughList::class, true)) {
+            $this->dataList = ManyManyThroughList::create(
+                $dataClass,
+                // If someone instantiates one of these and passes DataObjectSchema::manyManyComponent() directly
+                // the class will be in here as 'join'
+                $manyManyComponent['joinClass'] ?? $manyManyComponent['join'],
+                $manyManyComponent['childField'],
+                $manyManyComponent['parentField'],
+                $manyManyComponent['extraFields'],
+                $dataClass,
+                $manyManyComponent['parentClass']
+            );
+        } elseif (is_a($dataListClass, ManyManyList::class, true)) {
+            $this->dataList = ManyManyList::create(
+                $dataClass,
+                $manyManyComponent['join'],
+                $manyManyComponent['childField'],
+                $manyManyComponent['parentField'],
+                $manyManyComponent['extraFields']
+            );
+        } else {
+            $this->dataList = $dataListClass::create($dataClass, '');
+        }
+
+        if (isset($manyManyComponent['extraFields'])) {
+            $this->extraFields = $manyManyComponent['extraFields'];
+        }
+    }
+
+    /**
+     * Returns true if the variable passed in is valid for use in $this->dataList->forForeignID() on relation lists
+     */
+    private function isValidForeignID(int|array|null $foreignID): bool
+    {
+        // For an array, only return true if the array contains only integers and isn't empty
+        if (is_array($foreignID)) {
+            if (empty($foreignID)) {
+                return false;
+            }
+            foreach ($foreignID as $id) {
+                if (!is_int($id)) {
+                    return false;
+                }
+            }
+            return true;
+        }
+        // ID must be a valid ID int
+        return $foreignID !== null && $foreignID >= 1;
+    }
+
+    /**
+     * Pass in any eager-loaded data which applies to relations on a specific record in this list
+     *
+     * @return $this
+     */
+    public function addEagerLoadedData(string $relation, int $id, self|DataObject $data): static
+    {
+        $this->eagerLoadedData[$id][$relation] = $data;
+        return $this;
+    }
+
+    /**
+     * Get the dataClass name for this list, ie the DataObject ClassName
+     */
+    public function dataClass(): string
+    {
+        return $this->dataClass;
+    }
+
+    public function dbObject($fieldName): ?DBField
+    {
+        return singleton($this->dataClass)->dbObject($fieldName);
+    }
+
+    public function getIDList(): array
+    {
+        $ids = $this->column('ID');
+        return array_combine($ids, $ids) ?: [];
```

Comment:
```suggestion
        return array_combine($ids, $ids);
```
Is there any scenario where array_combine returns something falsey? https://www.php.net/manual/en/function.array-combine.php

Reply:
I feel like I did that based on something failing but I can't remember what, and I can't see how it _could_ return anything falsey (other than an empty array itself) so I'll remove that and if tests don't fail we'll call it good.

Reply:
Done